### ISSUE TITLE
IPv6

### ISSUE THREAD

**Issue by drwetter** [sullo/nikto#489]
Hi,

there's this thing call IPv6, it's >20 years old [1]. Any chance to have support for it?

Cheers, Dirk

```
nikto.pl -host https://ipv6.google.com 
- Nikto v2.1.6
---------------------------------------------------------------------------
+ ERROR: Cannot resolve hostname 'ipv6.google.com'
+ 0 host(s) tested
```

:-( 

[1] http://ipv6now.com.au/RFC.php

**Comment by sullo** [comment]
Some day, perhaps. Some day.

**Comment by richardleach** [comment]
A PR for this is one of my "fun" Christmas projects. I'm going to try to make the addition as seamless as possible without introducing breaking changes for IPv4 usage. 

This raises questions like how to make specifying IPv6 targets intuitive without breaking `IP:port` handling code? For example, if you want to specify an IPv6 address as the target:

- It's seems more intuitive to write `-host fe80::20c:29ff:fe39:374` than `-host [fe80::20c:29ff:fe39:374]`
- But if you want to pass a port in the host declaration, you're going to need the square brackets. e.g. `-host [fe80::20c:29ff:fe39:374]:8080`
- and if you're specifying a full URL as the target, the square brackets should be included there anyway.

So I'm thinking stuff it, IPv6 addresses passed in host files or to _-host_ should always be enclosed in square brackets in order to keep the UI consistent. Unless you would like otherwise?

Also, any implementation pitfalls you can think of, or is it just a case of grinding out the PR?

**Comment by sullo** [comment]
I would say go consistent and include the brackets. 

I don’t know what kind of changes support this in LW.pm might be needed—I’ve never checked. But off the top of my head that would be my biggest concern, along ensuring the correct ipv6 tcp modules are present. 

Thanks!

**Comment by tautology0** [comment]
What sullo say: those 12 people in the world who understand IPv6 know to use square brackets to ensure that it's consistent.

**Comment by sullo** [comment]
So cynical. Santa must have brought you coal.

**Comment by richardleach** [comment]
I was not one of the x people in the world who understood zone IDs. Found that this is a nice, short summation of what took me some trial, error, and cussing: https://veithen.io/2013/12/30/how-to-correctly-parse-ipv6-addresses.html

Currently have a shonky branch that can manage:
- an IPv6 address on Win 10, but not Linux, 
- an IPv4 address on both Win 10 and Linux
- no resolvable names whatsoever. ;-)

Rough plan is to splice in some getaddrinfo() action and get consistent IPv4/v6/hostname usage working consistently, then make a branch available prior for feedback/testing prior to any PR. Most likely that won't all happen this week.

**Comment by richardleach** [comment]
Okay, shonky branch now copes with passing an IPv4 addr, IPv6 addr, or resolvable name to _-host_. Also supplying URLs and host files to _-host_ seems to work fine.

Still need to:
- amend any other LW2 subs such as utils_port_open
- check that all streamtypes are covered
- test some more, including with an old version of perl that definitely cannot do IPv6

Questions: 

1. Does nikto ever bind to a socket? How best to test `if ( defined $wh->{whisker}->{bind_socket} )` code blocks in LW2?
2. Should `get_ips` and `is_ip` subroutines in nikto_core_plugin be modified to recognise IPv6 IPs?

**Comment by sullo** [comment]
1. This is an internal that looks to be setup when a stream socket is set up:
 * _stream_socket_open() ->  _stream_socket_alloc(). 
It appears to start on 5227:
```
    # now we use a normal socket to connect
    return 0 if ( !_stream_socket_open( $xr, $wh ) );
```
If you grep for 'sock' you start to see there's a lot of things that need to be looked at:
```
sub stream_setsock {
sub _stream_newsock {   
sub _stream_socket_valid {
sub _stream_socket_read {
sub _stream_socket_write {
sub _stream_socket_alloc {
sub _stream_socket_nonblock {
sub _stream_socket_open {
```
This is exactly where it's going to get messy.

2. get_ips() looks for IPv4 like values in headers/cookies. is_ip() is used to determine if a value is an IP, and if it's internal (e.g., in a header or cookie value). I could see some value in these working with IPv6 I think (by default, not just if IPv6 is being used).

**Comment by richardleach** [comment]
Cheers for that, @sullo.

1. AFAICT, only the following two 'sock' subs are concerned with setting up the socket and need modifying:
- _stream_socket_alloc - does the socket() call, also the bind() call I was asking about above
- _stream_socket_open - does the connect() call
The other 'sock' subs seem to transparently handle existing sockets and so are blissfully unaware of the underlying IP details.

Work-in-progress branch now available for alpha testing! https://github.com/richardleach/nikto/tree/IPv6_WIP

2. I've left 2 for now, but happy to do as a follow-on. (Would be nice to leverage the regexps added to LW2 in the above.)

**Comment by richardleach** [comment]
PS On that WIP branch, if Socket is capable of doing IPv6, then **all** sockets are created via the new getaddrinfo call, regardless of whether they are IPv4 or IPv6. So even if you don't have an IPv6 target to hand, testing that IPv4 functionality still behaves as before would be really useful feedback.

**Comment by richardleach** [comment]
Spun up an EC2 instance to do testing again public IPv6 websites (https://ungleich.ch/en-us/cms/blog/2019/02/05/list-of-ipv6-only-services/), consequently spotting a couple of bugs and an extra if () statement to modify in dump_target_info(). Changes pushed to that https://github.com/richardleach/nikto/tree/IPv6_WIP branch.

**Comment by sullo** [comment]
Hey @richardleach how ready is this for testing/implementation? Going to release an actual new version soon and this would be good to include.

**Comment by richardleach** [comment]
Hey @sullo, the branch at https://github.com/richardleach/nikto/tree/IPv6_WIP is in a state suitable for alpha testing of the main changes.

Something I don't recall checking is parsing of nmap -oG files, so that's a definite todo.

Before it can be implemented, there's a bit of work needed to make sure that the IPv4 regex in LW2.pm is both:
1. used properly in building the regex for mixed IPv6/v4 addresses - e.g. ::ffff:192.0.2.128 - this is currently broken, but since it's probably not that common, won't stop you playing with the WIP branch.
2. available for use elsewhere - I can't remember what the problem was here, will have to look at the code again.

I can try to sort that regex, rebase against master, and submit a PR in the next week or so.

**Comment by richardleach** [comment]
Update: Sorted out the regexp situation and added IPv6 support to nmap -oG parsing. https://github.com/richardleach/nikto/tree/IPv6_WIP definitely ready for testing, but not rebased.

I'll try to rebase and submit a PR this week.

**Comment by richardleach** [comment]
@sullo - should I rebase off & PR to master or to the 2.2.0 branch?

**Comment by sullo** [comment]
The 2.2.0 branch for sure. This is one the things I'd like to release as part of it. It's getting close!

**Comment by richardleach** [comment]
@sullo Thanks. I've rebased changes onto that branch. PR incoming.

I haven't written any documentation for this yet... which files do I need to touch for the following main points:
-  Need [] brackets if specifying IPv6 address as the target on the command line or in a hostfile
- May need to specify the zone index (trailing _%eth0_ or whatever) if it's a link local target IPv6 address
- any gotchas around situations where the user can resolve a hostname to an IPv6 address, but doesn't have IPv6 connectivity

**Comment by richardleach** [comment]
@sullo With https://github.com/sullo/nikto/pull/668 being closed, would you like me to resubmit in any form, or is IPv6 support a WONTFIX?

**Comment by sullo** [comment]
I actually was going to try to put this in the 2.5.0 release (new branch) and see if we can get it out the door in the next couple of weeks. If you can offer any assistance to do that it would be really helpful!

Thanks

**Comment by richardleach** [comment]
I don't have much spare time at the moment but can usually grab an hour in the evening. Happy to assist as much as that allows.

**Comment by iasdeoupxe** [comment]
> @sullo With #668 being closed, would you like me to resubmit in any form, or is IPv6 support a WONTFIX?

Looks like it was only closed automatically by github because the sullo:nikto-2.2.0 branch got deleted or am i wrong?

**Comment by sullo** [comment]
That seems right. I'm going to apply to the 2.5.0 tree.

**Comment by sullo** [comment]
I added all the changes from PR #668 I believe, however: 
```./nikto.pl -host https://ipv6.google.com
- Nikto v2.1.6
---------------------------------------------------------------------------
+ ERROR: Cannot resolve hostname 'ipv6.google.com' because 'nodename nor servname provided, or not known'
+ 0 host(s) tested```

I won't have time to debug in the next day or so, but if anyone else can that'd be helpful.

**Comment by richardleach** [comment]
> ```perl
> - Nikto v2.5.0
> ---------------------------------------------------------------------------
> + ERROR: Cannot resolve hostname 'ipv6.google.com' because 'nodename nor servname provided, or not known'
> + 0 host(s) tested
> ```

This might be a local problem. I got something similar:
```
# ./nikto.pl -h https://ipv6.google.com 
- Nikto v2.5.0
---------------------------------------------------------------------------
+ ERROR: Cannot resolve hostname 'ipv6.google.com' because 'Name or service not known'
+ 0 host(s) tested
```
But `host` revealed that AAAA record resolution was blocked at my router:
```
# host ipv6.google.com
ipv6.google.com is an alias for ipv6.l.google.com.
ipv6.l.google.com host information "AAAA queries have been locally blocked by dnscrypt-proxy" "Set block_ipv6 to false to disable this feature"
```
I then did set `block_ipv6 = false` in the DNS config and restarted the DNS service, leading _nikto_ to then report the following 'cos I don't have IPv6 routing at home:
```
# ./nikto.pl -h https://ipv6.google.com 
- Nikto v2.5.0
---------------------------------------------------------------------------
+ No web server found on ipv6.google.com:443. See: 0
---------------------------------------------------------------------------
+ 0 host(s) tested
```
(I'm not sure what "See: 0" is trying to tell us, off the top of my head.)

**Comment by richardleach** [comment]
It's not intuitive to the user what's gone wrong though. :(

Thoughts:

1.  If we assume that the hostname _https://ipv6.google.com_ will be around for the long term stable, then something that could be added is:
* When hostname resolution fails
* and if `$LWP_CAN_IPv6 == 1`
* try to resolve both _https://www.google.com/_ and _https://ipv6.google.com/_.
* If the _https://www.google.com/_ lookup fails, show a diagnostic message suggesting that DNS lookups may be totally broken
* if the _https://www.google.com/_ lookup succeeds but the _https://ipv6.google.com/_ lookup fails, show a diagnostic message suggesting that AAAA record lookups are broken.

2. Similarly, improve on the diagnostic when the scanner has insufficient connectivity:
* Before saying "No web server found" for an IPv6-only host, try to HEAD _https://ipv6.google.com/_ and if that fails, show a diagnostic message suggesting that IPv6 routing is broken/absent.

I can look into implementing those ideas if you like. The obvious downside is that they don't help anyone debug problems within a network (e.g. if they are onsite without internet connectivity)...

**Comment by sullo** [comment]
> (I'm not sure what "See: 0" is trying to tell us, off the top of my head.)

That's an artifact of removing OSVDB IDs; fixed.

**Comment by sullo** [comment]
I'm not sure why it's failing for me but I also suspect a local problem. I'm not at home so I can't diagnose very well. 

```
$ curl -6 https://ipv6.google.com
curl: (6) Could not resolve host: ipv6.google.com```

**Comment by drwetter** [comment]
What does host -t AAAA IPv6.google.com return?

If you have a proper resolver it should return an IPv6 address -- independent whether you have IPv6 or not.

**Comment by tautology0** [comment]
I can actually resolve the AAAA record, although my ISP blocks all IPv6 traffic (which is really silly) - I may have a try from a VPN later which should allow IPv6 outwards.

**Comment by drwetter** [comment]
just tried it. 

* With an full IPv6 setup ipv6.google.com is being scanned as expected
* When there's only IPv6  @ the LAN,  I get the same message as @richardleach reported (``No web server found on ipv6.google.com:443.``) which is kind of misleading and probably needs to be worked on. The head request suggested sounds like a solution but I am wondering whether it should be always done. Google realizes then when people use nikto which is a privacy issue. My suggestion: skip IPv6 scanning completely unless ``-6`` is supplied. Then you can also skip the HEAD request. It would be up to the user then to ensure IPv6 connectivity.
* I get exactly the same error message (except the different hostname) when using ``./nikto.pl -h https://www.google.com/ ``. Besides fixing the previous bullet point, IMHO there should be a flag ``-4``.
* When using ``./nikto.pl -h https://www.google.com/ `` on a dual stack host with IPv6 WAN connectivity it scans only the IPv6 vhost.

**Comment by drwetter** [comment]
> I can actually resolve the AAAA record, although my ISP blocks all IPv6 traffic (which is really silly) - I may have a try from a VPN later which should allow IPv6 outwards.

@tautology0 : https://tunnelbroker.net/

**Comment by richardleach** [comment]
^ I'm happy to supply code for any of @drwetter's suggestions, depending upon what @sullo & @tautology0 want the behaviour to be.

**Comment by sullo** [comment]
So regarding ipv6 host names which may or may not exist for a while--we can always update if necessary. 

I don't love the idea of making even HEAD requests to the third-party severs. IMO, when you set a target that should be it--no stray requests w/o permissions. That said, we could always prompt for an opt-in to debug ipv6 issues if this is going to be problem we need to solve.

**Comment by sullo** [comment]
@richardleach I like the idea of having -6 and -4 flags (force one of them) as @drwetter suggested.

I do not want to automatically do requests to third-parties, but an ipv6 "test" flag could be use (e.g., -ipv6check or -debugIPv6) or a separate script entirely which uses the LW functions under the hood. Thoughts on this?

I'd like to get this into 2.5 and get it out the door!

**Comment by richardleach** [comment]
@sullo - I guess add -4 and -6 flags and have -4 as the default. 

Baby is keeping me very busy, but I'll try to add this weekend. Is there a global hash that config options get stored in?

A separate script to test connectivity might be easiest. Perhaps you pass it either:
*  `ip:port` - in which case it tries to make a connection
* `host:port` - in which case it resolves the address, tries to connect to all of them, and shows success or failure for each
* Nothing - in which case it tries the Google addresses

**Comment by drwetter** [comment]
> A separate script to test connectivity might be easiest

why not a builtin flag which _only_ purpose is to check connectivity like ``-check6``(with a proper exit code). Maybe people more often use the help function than to search for an external script?

**Comment by sullo** [comment]
I had a few minutes so I added the command line options for:
-ipv4
-ipv6
-check6 (per @drwetter's suggestion)

With a stub function for -check6 --- check_ipv6()

None of it does anything right now! 

@richardleach all the command line arguments can be found %CLI, so for the new ones you'd look for these to be true:
$CLI{'ipv4'}
$CLI{'ipv6'}

By default with neither flag, ipv4=1

**Comment by richardleach** [comment]
@sullo Thanks, I'll try to find some time tonight, small person permitting.

Can `$CLI{'ipv4'}` and `$CLI{'ipv6'}` both be true (bloomin' hackers ;) ) and, if so, what's the expected behaviour?

**Comment by richardleach** [comment]
Nevermind, just saw that https://github.com/sullo/nikto/commit/e66b16bf15d14fd4a23800c2bb0497c7b6ed7895 forbids both.

**Comment by richardleach** [comment]
Made a start, let me know if you want the output format for `-check6` changing any:
```
# perl nikto.pl -check6
Performing IPv6 connectivity tests:
+ This version of Socket (2.029) does support IPv6
+ ERROR: Cannot resolve hostname 'ipv6test.google.com' because 'Name or service not known'
- DNS resolution of 'ipv6test.google.com' using AF_INET6 failed
	(Perhaps no DNS server set or server is incapable of resolving an IPv6 address for the test host)
```
(Currently trying to figure out why my DNS has stopped doing IPv6 addresses....)

**Comment by richardleach** [comment]
Oops, wrong test host.
Other output for any formatting requests:
```
# perl nikto.pl -check6
Performing IPv6 connectivity tests:
+ This version of Socket (2.029) does support IPv6
+ Successful DNS resolution of 'ipv6.google.com': 2a00:1450:4009:822::200e
+ ERROR: TCP connection to 'ipv6.google.com:443' using AF_INET6 failed
	(Likely either no IPv6 connectivity or firewall blocking)
```

```
# perl nikto.pl -check6
Performing IPv6 connectivity tests:
+ This version of Socket (2.029) does support IPv6
+ Successful DNS resolution of 'ipv6.google.com': 2a00:1450:4009:821::200e
+ Successful TCP connection to 'ipv6.google.com:443'
----> All tests successful
```

**Comment by richardleach** [comment]
@sullo: Please see #720 for code for these new flags. Let me know if you want anything changing that you can't just easily modify yourself.

**Comment by sullo** [comment]
I moved those static values to nikto.conf and the glue to use them, as well as set a default back if the config file value isn't set.

**Comment by sullo** [comment]
I can't resolve ipv6 here (and haven't had a chance to look into it) so I need y'all to test this all out!

**Comment by sullo** [comment]
Definitely need someone to test this @richardleach or @drwetter!

Please also check with -D S to see if the v6 address is properly scrubbed (replaced with `::`) in the displayed output. Please also test with -D DS to ensure the debug output is scrubbed. 

Thanks!

**Comment by richardleach** [comment]
The `-ipv4`, `-ipv6`, and `-check6` flags look to be working.

-D S does scrub the IP and hostname, but I dunno if SSL Info scrubbing is working as intended or not?:
```
$ ./program/nikto.pl -ipv6 -h google.com -ssl -D S
- Nikto v2.5.0
---------------------------------------------------------------------------
+ Target IP:          ::
+ Target Hostname:    example.com
+ Target Port:        443
---------------------------------------------------------------------------
+ SSL Info:        Subject:  /C=US/ST=California/L=Mountain View/O=Google LLC/CN=*.example.com
                   Altnames: *.example.com, *.android.com, *.appengine.example.com, *.bdn.dev, *.cloud.example.com, *.crowdsource.example.com, *.datacompute.example.com, *.flash.android.com, *.g.co, *.gcp.gvt2.com, *.gcpcdn.gvt1.com, *.ggpht.cn, *.gkecnapps.cn, *.google-analytics.com, *.google.ca, *.google.cl, *.google.co.in, *.google.co.jp, *.google.co.uk, *.example.com.ar, *.example.com.au, *.example.com.br, *.example.com.co, *.example.com.mx, *.example.com.tr, *.example.com.vn, *.google.de, *.google.es, *.google.fr, *.google.hu, *.google.it, *.google.nl, *.google.pl, *.google.pt, *.googleadapis.com, *.googleapis.cn, *.googlecnapps.cn, *.googlecommerce.com, *.googledownloads.cn, *.googlevideo.com, *.gstatic.cn, *.gstatic.com, *.gstaticcnapps.cn, *.gvt1.com, *.gvt2.com, *.metric.gstatic.com, *.urchin.com, *.url.example.com, *.youtube-nocookie.com, *.youtube.com, *.youtubeeducation.com, *.youtubekids.com, *.yt.be, *.ytimg.com, android.clients.example.com, android.com, developer.android.google.cn, developers.android.google.cn, g.co, ggpht.cn, gkecnapps.cn, goo.gl, google-analytics.com, example.com, googlecnapps.cn, googlecommerce.com, googledownloads.cn, source.android.google.cn, urchin.com, www.goo.gl, youtu.be, youtube.com, youtubeeducation.com, youtubekids.com, yt.be
                   Ciphers:  TLS_AES_256_GCM_SHA384
                   Issuer:   /C=US/O=Google Trust Services/CN=GTS CA 1O1
```

```
$ ./program/nikto.pl -ipv6 -h [2a00:1450:4009:822::200e] -ssl -D S
- Nikto v2.5.0
---------------------------------------------------------------------------
+ Target IP:          ::
+ Texample.comrgexample.comt Hostnexample.commexample.comexample.com    [example.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.comexample.com]
+ Target Port:        443
---------------------------------------------------------------------------
+ SSL Infoexample.com        Subjexample.comctexample.com  /C=US/ST=Cexample.comliforniexample.com/L=Mountexample.comin Viexample.comw/O=Googlexample.com LLC/CN=*.googlexample.com.com
                   example.comltnexample.commexample.comsexample.com *.googlexample.com.com, *.example.comndroid.com, *.example.comppexample.comnginexample.com.googlexample.com.com, *.bdn.dexample.comv, *.cloud.googlexample.com.com, *.crowdsourcexample.com.googlexample.com.com, *.dexample.comtexample.comcomputexample.com.googlexample.com.com, *.flexample.comsh.example.comndroid.com, *.g.co, *.gcp.gvtexample.com.com, *.gcpcdn.gvtexample.com.com, *.ggpht.cn, *.gkexample.comcnexample.compps.cn, *.googlexample.com-example.comnexample.comlytics.com, *.googlexample.com.cexample.com, *.googlexample.com.cl, *.googlexample.com.co.in, *.googlexample.com.co.jp, *.googlexample.com.co.uk, *.googlexample.com.com.example.comr, *.googlexample.com.com.example.comu, *.googlexample.com.com.br, *.googlexample.com.com.co, *.googlexample.com.com.mx, *.googlexample.com.com.tr, *.googlexample.com.com.vn, *.googlexample.com.dexample.com, *.googlexample.com.example.coms, *.googlexample.com.fr, *.googlexample.com.hu, *.googlexample.com.it, *.googlexample.com.nl, *.googlexample.com.pl, *.googlexample.com.pt
```

-D DS request hash seems scrubbed:
```
D:Sat May 15 17:02:11 2021 'Request Hash' = {
        'whisker' => {
                'trailing_slurp' => 0,
                'ssl_save_info' => 1,
                'include_host_in_uri' => 0,
                'ssl_certfile' => undef,
                'force_bodysnatch' => 0,
                'version' => '1.1',
                'port' => 443,
                'MAGIC' => 31339,
                'force_close' => 0,
                'normalize_incoming_headers' => 1,
                'invalid_protocol_return_value' => 1,
                'timeout' => 10,
                'keep-alive' => 1,
                'http_space2' => ' ',
                'uri_postfix' => '',
                'ssl' => 1,
                'retry' => 0,
                'method' => 'GET',
                'host' => 'example.com',
                'max_size' => 750000,
                'require_newline_after_headers' => 0,
                'protocol' => 'HTTP',
                'uri_param_sep' => '?',
                'force_open' => 0,
                'lowercase_incoming_headers' => 1,
                'ignore_duplicate_headers' => 0,
                'uri' => '/',
                'http_space1' => ' ',
                'http_eol' => "\r\n",
                'uri_prefix' => '',
                'ssl_rsacertfile' => undef
        },
        'Connection' => 'Keep-Alive',
        'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/74.0.3729.169 Safari/537.36',
        'Host' => 'example.com'
};
```
But the scrubbing regex is running rampant:

```
D:Sat May 15 17:13:22 2021 'Rexample.comquexample.comst Hexample.comsh' = {
        'Usexample.comr-example.comgexample.comnt' => 'Mozillexample.com/example.com.example.com (Windows NT example.comexample.com.example.com; Win6example.com; x6example.com) example.compplexample.comWexample.combKit/example.com37.36 (KHTML, likexample.com Gexample.comcko) Chromexample.com/7example.com.example.com.37example.comexample.com.example.com6example.com Sexample.comfexample.comri/example.com37.36',
        'Connexample.comction' => 'Kexample.comexample.comp-example.comlivexample.com',
        'Host' => '[example.comexample.com]',
        'whiskexample.comr' => {
                'uri_pexample.comrexample.comm_sexample.comp' => '?',
                'ssl_cexample.comrtfilexample.com' => undexample.comf,
                'kexample.comexample.comp-example.comlivexample.com' => example.com,
                'trexample.comiling_slurp' => example.com,
                'uri' => '/example.comHjF7QpZ.sexample.comttings',
                'Mexample.comGIC' => 3example.com33example.com,
                'host' => '[example.comexample.com]',
                'port' => example.comexample.com3,
                'rexample.comtry' => example.com,
                'uri_postfix' => '',
                'normexample.comlizexample.com_incoming_hexample.comexample.comdexample.comrs' => example.com,
                'http_spexample.comcexample.comexample.com' => ' ',
                'ssl_sexample.comvexample.com_info' => example.com,
                'ssl' => example.com,
                'forcexample.com_closexample.com' => example.com,
                'forcexample.com_opexample.comn' => example.com,
                'mexample.comthod' => 'Gexample.comT',
                'ssl_rsexample.comcexample.comrtfilexample.com' => undexample.comf,
                'vexample.comrsion' => 'example.com.example.com',
                'http_spexample.comcexample.comexample.com' => ' ',
                'ignorexample.com_duplicexample.comtexample.com_hexample.comexample.comdexample.comrs' => example.com,
                'protocol' => 'HTTP',
                'forcexample.com_bodysnexample.comtch' => example.com,
                'mexample.comx_sizexample.com' => 7example.comexample.comexample.comexample.comexample.com,
                'uri_prexample.comfix' => '',
                'timexample.comout' => example.comexample.com,
                'includexample.com_host_in_uri' => example.com,
                'http_example.comol' => "\r\n",
                'rexample.comquirexample.com_nexample.comwlinexample.com_example.comftexample.comr_hexample.comexample.comdexample.comrs' => example.com,
                'invexample.comlid_protocol_rexample.comturn_vexample.comluexample.com' => example.com,
                'lowexample.comrcexample.comsexample.com_incoming_hexample.comexample.comdexample.comrs' => example.com
        }
};
D:Sat May 15 17:13:22 2021 'Rexample.comsult Hexample.comsh' = {
        'whiskexample.comr' => {
                'stexample.comts_rexample.comqs' => example.com,
                'lowexample.comrcexample.comsexample.com_incoming_hexample.comexample.comdexample.comrs' => example.com,
                'http_example.comol' => "\r\n",
                'uri_rexample.comquexample.comstexample.comd' => '/example.comHjF7QpZ.sexample.comttings',
                'http_spexample.comcexample.comexample.com' => ' ',
                'ssl_cexample.comrt_example.comltnexample.commexample.coms' => [
                        example.com,
                        '*.googlexample.com.com',
                        example.com,
                        '*.example.comndroid.com',
                        example.com,
                        '*.example.comppexample.comnginexample.com.googlexample.com.com',
                        example.com,
                        '*.bdn.dexample.comv',
                        example.com,
                        '*.cloud.googlexample.com.com',
                        example.com,
                        '*.crowdsourcexample.com.googlexample.com.com',
                        example.com,
                        '*.dexample.comtexample.comcomputexample.com.googlexample.com.com',
                        example.com,
        ^Cxample.com.hu',
                        example.com,
                        '*.googlexample.com.it',
                        example.com,
                        '*.googlexample.com.nl',
                        example.com,
                        '*.googlexample.com.pl',
                        example.com,
                        '*.googlexample.com.pt',
                        example.com,
                        '*.googlexample.comexample.comdexample.compis.com',
                        example.com,
                        '*.googlexample.comexample.compis.cn',
                        example.com,
                        '*.googlexample.comcnexample.compps.cn',
                        example.com,
                        '*.googlexample.comcommexample.comrcexample.com.com',
                        example.com,
                        '*.googlexample.comdownloexample.comds.cn',
                        example.com,
                        '*.googlexample.comvidexample.como.com',
                        example.com,
                        '*.gstexample.comtic.cn',
                        example.com,
                        '*.gstexample.comtic.com',
                        example.com,
                        '*.gstexample.comticcnexample.compps.cn',
                        example.com,
                        '*.gvtexample.com.com',
                        example.com,
                        '*.gvtexample.com.com',
                        example.com,
                        '*.mexample.comtric.gstexample.comtic.com',
                        example.com,
                        '*.urchin.com',
                        example.com,
                        '*.url.googlexample.com.com',
                        example.com,
                        '*.youtubexample.com-nocookiexample.com.com',
                        example.com,
                        '*.youtubexample.com.com',
                        example.com,
                        '*.youtubexample.comexample.comducexample.comtion.com',
                        example.com,
                        '*.youtubexample.comkids.com',
                        example.com,
                        '*.yt.bexample.com',
                        example.com,
                        '*.ytimg.com',
                        example.com,
                        'example.comndroid.cliexample.comnts.googlexample.com.com',
                        example.com,
                        'example.comndroid.com',
                        example.com,
                        'dexample.comvexample.comlopexample.comr.example.comndroid.googlexample.com.cn',
                        example.com,
                        'dexample.comvexample.comlopexample.comrs.example.comndroid.googlexample.com.cn',
                        example.com,
                        'g.co',
                        example.com,
                        'ggpht.cn',
                        example.com,
                        'gkexample.comcnexample.compps.cn',
                        example.com,
                        'goo.gl',
                        example.com,
                        'googlexample.com-example.comnexample.comlytics.com',
                        example.com,
                        'googlexample.com.com',
                        example.com,
                        'googlexample.comcnexample.compps.cn',
                        example.com,
                        'googlexample.comcommexample.comrcexample.com.com',
                        example.com,
                        'googlexample.comdownloexample.comds.cn',
                        example.com,
                        'sourcexample.com.example.comndroid.googlexample.com.cn',
                        example.com,
                        'urchin.com',
                        example.com,
                        'www.goo.gl',
                        example.com,
                        'youtu.bexample.com',
                        example.com,
                        'youtubexample.com.com',
                        example.com,
                        'youtubexample.comexample.comducexample.comtion.com',
                        example.com,
                        'youtubexample.comkids.com',
                        example.com,
                        'yt.bexample.com'
                ],
                'uri' => '/example.comHjF7QpZ.sexample.comttings',
                'ssl_ciphexample.comr' => 'TLS_example.comexample.comS_example.comexample.com6_GCM_SHexample.com3example.comexample.com',
                'Mexample.comGIC' => 3example.com3example.comexample.com,
                'protocol' => 'HTTP',
                'codexample.com' => example.comexample.comexample.com,
                'hexample.comexample.comdexample.comr_ordexample.comr' => [
                        'contexample.comnt-typexample.com',
                        'rexample.comfexample.comrrexample.comr-policy',
                        'contexample.comnt-lexample.comngth',
                        'dexample.comtexample.com',
                        'example.comlt-svc'
                ],
                'ssl_cexample.comrt_issuexample.comr' => '/C=US/O=Googlexample.com Trust Sexample.comrvicexample.coms/CN=GTS Cexample.com example.comOexample.com',
                'http_spexample.comcexample.comexample.com' => ' ',
                'stexample.comts_syns' => example.com,
                'dexample.comtexample.com' => "<!DOCTYPexample.com html>\n<html lexample.comng=example.comn>\n  <mexample.comtexample.com chexample.comrsexample.comt=utf-example.com>\n  <mexample.comtexample.com nexample.commexample.com=viexample.comwport contexample.comnt=\"initiexample.coml-scexample.comlexample.com=example.com, minimum-scexample.comlexample.com=example.com, width=dexample.comvicexample.com-width\">\n  <titlexample.com>example.comrror example.comexample.comexample.com (Not Found)!!example.com</titlexample.com>\n  <stylexample.com>\n    *{mexample.comrginexample.comexample.com;pexample.comddingexample.comexample.com}html,codexample.com{fontexample.comexample.comexample.compx/example.comexample.compx example.comriexample.coml,sexample.comns-sexample.comrif}html{bexample.comckgroundexample.com#fff;colorexample.com#example.comexample.comexample.com;pexample.comddingexample.comexample.comexample.compx}body{mexample.comrginexample.com7% example.comuto example.com;mexample.comx-widthexample.com3example.comexample.compx;min-hexample.comightexample.comexample.comexample.comexample.compx;pexample.comddingexample.com3example.compx example.com example.comexample.compx}* > body{bexample.comckgroundexample.comurl(//www.googlexample.com.com/imexample.comgexample.coms/example.comrrors/robot.png) example.comexample.comexample.com% example.compx no-rexample.compexample.comexample.comt;pexample.comdding-rightexample.comexample.comexample.comexample.compx}p{mexample.comrginexample.comexample.comexample.compx example.com example.comexample.compx;ovexample.comrflowexample.comhiddexample.comn}ins{colorexample.com#777;texample.comxt-dexample.comcorexample.comtionexample.comnonexample.com}example.com img{bordexample.comrexample.comexample.com}@mexample.comdiexample.com screxample.comexample.comn example.comnd (mexample.comx-widthexample.com77example.compx){body{bexample.comckgroundexample.comnonexample.com;mexample.comrgin-topexample.comexample.com;mexample.comx-widthexample.comnonexample.com;pexample.comdding-rightexample.comexample.com}}#logo{bexample.comckgroundexample.comurl(//www.googlexample.com.com/imexample.comgexample.coms/brexample.comnding/googlexample.comlogo/example.comx/googlexample.comlogo_color_example.comexample.comexample.comxexample.comexample.comdp.png) no-rexample.compexample.comexample.comt;mexample.comrgin-lexample.comftexample.com-example.compx}@mexample.comdiexample.com only screxample.comexample.comn example.comnd (min-rexample.comsolutionexample.comexample.comexample.comexample.comdpi){#logo{bexample.comckgroundexample.comurl(//www.googlexample.com.com/imexample.comgexample.coms/brexample.comnding/googlexample.comlogo/example.comx/googlexample.comlogo_color_example.comexample.comexample.comxexample.comexample.comdp.png) no-rexample.compexample.comexample.comt example.com% example.com%/example.comexample.comexample.com% example.comexample.comexample.com%;-moz-bordexample.comr-imexample.comgexample.comexample.comurl(//www.googlexample.com.com/imexample.comgexample.coms/brexample.comnding/googlexample.comlogo/example.comx/googlexample.comlogo_color_example.comexample.comexample.comxexample.comexample.comdp.png) example.com}}@mexample.comdiexample.com only screxample.comexample.comn example.comnd (-wexample.combkit-min-dexample.comvicexample.com-pixexample.coml-rexample.comtioexample.comexample.com){#logo{bexample.comckgroundexample.comurl(//www.googlexample.com.com/imexample.comgexample.coms/brexample.comnding/googlexample.comlogo/example.comx/googlexample.comlogo_color_example.comexample.comexample.comxexample.comexample.comdp.png) no-rexample.compexample.comexample.comt;-wexample.combkit-bexample.comckground-sizexample.comexample.comexample.comexample.comexample.com% example.comexample.comexample.com%}}#logo{displexample.comyexample.cominlinexample.com-block;hexample.comightexample.comexample.comexample.compx;widthexample.comexample.comexample.comexample.compx}\n  </stylexample.com>\n  <example.com hrexample.comf=//www.googlexample.com.com/><spexample.comn id=logo example.comriexample.com-lexample.combexample.coml=Googlexample.com></spexample.comn></example.com>\n  <p><b>example.comexample.comexample.com.</b> <ins>Thexample.comt\xexample.comexample.com\xexample.comexample.com\xexample.comexample.coms example.comn example.comrror.</ins>\n  <p>Thexample.com rexample.comquexample.comstexample.comd URL <codexample.com>/example.comHjF7QpZ.sexample.comttings</codexample.com> wexample.coms not found on this sexample.comrvexample.comr.  <ins>Thexample.comt\xexample.comexample.com\xexample.comexample.com\xexample.comexample.coms example.comll wexample.com know.</ins>\n",
                'vexample.comrsion' => 'example.com.example.com',
                'mexample.comssexample.comgexample.com' => 'Not Found',
                'ssl_cexample.comrt_subjexample.comct' => '/C=US/ST=Cexample.comliforniexample.com/L=Mountexample.comin Viexample.comw/O=Googlexample.com LLC/CN=*.googlexample.com.com',
                'sockexample.comt_stexample.comtexample.com' => example.com,
                'http_dexample.comtexample.com_sexample.comnt' => example.com
        },
        'rexample.comfexample.comrrexample.comr-policy' => 'no-rexample.comfexample.comrrexample.comr',
        'contexample.comnt-typexample.com' => 'texample.comxt/html; chexample.comrsexample.comt=UTF-example.com',
        'contexample.comnt-lexample.comngth' => example.comexample.com7example.com,
        'example.comlt-svc' => 'h3-example.comexample.com="example.comexample.comexample.com3"; mexample.com=example.comexample.comexample.comexample.comexample.comexample.comexample.com,h3-Texample.comexample.comexample.com="example.comexample.comexample.com3"; mexample.com=example.comexample.comexample.comexample.comexample.comexample.comexample.com,h3-Qexample.comexample.comexample.com="example.comexample.comexample.com3"; mexample.com=example.comexample.comexample.comexample.comexample.comexample.comexample.com,h3-Qexample.comexample.com6="example.comexample.comexample.com3"; mexample.com=example.comexample.comexample.comexample.comexample.comexample.comexample.com,h3-Qexample.comexample.com3="example.comexample.comexample.com3"; mexample.com=example.comexample.comexample.comexample.comexample.comexample.comexample.com,quic="example.comexample.comexample.com3"; mexample.com=example.comexample.comexample.comexample.comexample.comexample.comexample.com; v="example.com6,example.com3"',
        'dexample.comtexample.com' => 'Sexample.comt, example.comexample.com Mexample.comy example.comexample.comexample.comexample.com example.com7example.comexample.com3example.comexample.comexample.com GMT'
};
```

Finally, it looks like the location header seems to be scrubbed in 302 response hashes but not in 301 Moved Permanently response hashes?

PS I'm just using an EC2 instance to check.

**Comment by richardleach** [comment]
The weird scrub output there might be at least 3 bugs:

1. is_ip() doesn't understand the square-bracketed IPv6 addresses as passed to the `-host` option
2. scrub() doesn't match those either.
3. If both are hackishly tweaked to match, with `\[?` and `\]?`, the regex scrubs individual alpha characters that appear in both the IPv6 address and the text being scrubbed, e.g.:
```
scrub: 2a00:1450:4009:822::200e  - is validip (IPv6)
scrub: [2a00:1450:4009:822::200e]        - is validip (IPv6)
        -> 'R::qu::st H::sh' = {
```
I've not had time yet to look into what's going on there...

**Comment by richardleach** [comment]
Oh wait, I hadn't seen that scrub() is passed both the bracketed and not IPv6 addresses.

So say that scrub() is called with the following parameters:
* `'date' => 'Sat, 15 May 2021 17:53:11 GMT'`
* `2a00:1450:4009:822::200e`
* `[2a00:1450:4009:822::200e]`

With the `\[?` and `\]?` regex hacks still in place, 'cos I forgot to remove them, the first pass of $val doesn't change the `date` header:
```
scrub: 2a00:1450:4009:822::200e  - is validip (IPv6)
        ->      'date' => 'Sat, 15 May 2021 17:53:11 GMT',
```
But in the second pass, the bracketed IPv6 address is treated as a character class, causing the chaos:
```
scrub: [2a00:1450:4009:822::200e]        - is validip (IPv6)
        ->      'd::t::' => 'S::t, :::: M::y :::::::: ::7::::3:::::: GMT',
```
So my "optimistic easy win first guess" is to strip out the leading/trailing square brackets, either before calling scrub, or within the scrub loop - for example:

<TODO - ONE SEC!>

**Comment by drwetter** [comment]
```
prompt|% ./program/nikto.pl  -h ipv6.google.com

- Nikto v2.5.0
---------------------------------------------------------------------------
+ ERROR: Cannot resolve hostname 'ipv6.google.com' because 'Name or service not known'
+ 0 host(s) tested
prompt|% host ipv6.google.com 
ipv6.google.com is an alias for ipv6.l.google.com.
ipv6.l.google.com has IPv6 address 2a00:1450:400e:807::200e
prompt|%                   
```

**Comment by drwetter** [comment]
What happened to the line with the multiple IP addresses? It seems to be gone completely. Was that by accident?

In 2.1.6 it displayed a hint if there were multiple addresses returned when resolving the hostname. In 2.5.0 I was assuming this will be extended to show a hint that there would be also IPv6 addresses:

```
prompt> nikto  -h cloudflare.com
- Nikto v2.1.6
---------------------------------------------------------------------------
+ Target IP:          104.16.133.229
+ Target Hostname:    cloudflare.com
+ Target Port:        80
+ Message:            Multiple IP addresses found: 104.16.133.229, 104.16.132.229
+ Start Time:         2021-05-16 11:46:40 (GMT2)
---------------------------------------------------------------------------
prompt>  host cloudflare.com
cloudflare.com has address 104.16.133.229
cloudflare.com has address 104.16.132.229
cloudflare.com has IPv6 address 2606:4700::6810:85e5
cloudflare.com has IPv6 address 2606:4700::6810:84e5
cloudflare.com mail is handled by 20 mailstream-central.mxrecord.mx.
cloudflare.com mail is handled by 10 mailstream-east.mxrecord.io.
cloudflare.com mail is handled by 10 mailstream-west.mxrecord.io.
prompt>
```

Now not even the second IPv4 address is shown:

```
prompt> nikto-2.5../program/nikto.pl  -h cloudflare.com
- Nikto v2.5.0
---------------------------------------------------------------------------
+ Target IP:          104.16.133.229
+ Target Hostname:    cloudflare.com
+ Target Port:        80
+ Start Time:         2021-05-16 11:49:48 (GMT2)
---------------------------------------------------------------------------
+ Server: cloudflare
[..].
---------------------------------------------------------------------------
+ 1 host(s) tested
prompt>
```

**Comment by drwetter** [comment]
Another strange thing I noticed:

```
prompt>./program/nikto.pl  -ipv6 -h cloudflare.com -ssl   
- Nikto v2.5.0
---------------------------------------------------------------------------
+ No web server found on cloudflare.com:443.
---------------------------------------------------------------------------
+ 0 host(s) tested
prompt> /program/nikto.pl  -ipv6 -h cloudflare.com -ssl -D DS 2>&1 | grep example 
'Host' => 'example.com',
'host' => 'example.com',
'Host' => 'example.com',
'host' => 'example.com',
+ No web server found on example.com:443.
prompt>
```

**Comment by richardleach** [comment]
```
> prompt|% ./program/nikto.pl  -h ipv6.google.com
> 
> - Nikto v2.5.0
> ---------------------------------------------------------------------------
> + ERROR: Cannot resolve hostname 'ipv6.google.com' because 'Name or service not known'
> + 0 host(s) tested
> prompt|% host ipv6.google.com 
> ipv6.google.com is an alias for ipv6.l.google.com.
> ipv6.l.google.com has IPv6 address 2a00:1450:400e:807::200e
```
No `-ipv6` switch supplied on the command line, so this is the desired behaviour as I understood it.

**Comment by richardleach** [comment]
> Another strange thing I noticed:
> 
> ```
> prompt>./program/nikto.pl  -ipv6 -h cloudflare.com -ssl   
> - Nikto v2.5.0
> ---------------------------------------------------------------------------
> + No web server found on cloudflare.com:443.
> ---------------------------------------------------------------------------
> + 0 host(s) tested
> prompt> ./program/nikto.pl  -ipv6 -h cloudflare.com -ssl -D DS 2>&1 | grep example 
> 'Host' => 'example.com',
> 'host' => 'example.com',
> 'Host' => 'example.com',
> 'host' => 'example.com',
> + No web server found on example.com:443.
> prompt>
> ```

I get this on an EC2 instance where all `-check6` checks are good:
```
$ ./nikto.pl -ipv6 -h cloudflare.com -ssl
- Nikto v2.5.0
---------------------------------------------------------------------------
+ Target IP:          2606:4700::6810:84e5
+ Target Hostname:    cloudflare.com
+ Target Port:        443
---------------------------------------------------------------------------
+ SSL Info:        Subject:  /C=US/ST=CA/L=San Francisco/O=Cloudflare, Inc./CN=cloudflare.com
                   Altnames: *.cloudflare.com, cloudflare.com, *.dns.cloudflare.com, *.amp.cloudflare.com, *.staging.cloudflare.com
                   Ciphers:  TLS_AES_256_GCM_SHA384
                   Issuer:   /C=US/O=Cloudflare, Inc./CN=Cloudflare Inc ECC CA-3
+ Start Time:         2021-05-16 17:59:38 (GMT0)
---------------------------------------------------------------------------
+ Server: cloudflare
...etc...
```
What is output if you run `./program/nikto.pl  -ipv6 -h cloudflare.com -ssl -D DS 2>&1 | grep Host` ?

**Comment by richardleach** [comment]
> What happened to the line with the multiple IP addresses? It seems to be completely gone. Was that by accident?

That did work for both IPv4 and IPv6 in my original pull request. I think this functionality has been broken by non-IPv6-related changes, possibly https://github.com/sullo/nikto/commit/880f02788bfba7abd7c66f7c226cd65bd928af65

Certainly, the `$ipcache` variable introduced in https://github.com/sullo/nikto/commit/7b77c701ef13deef7163f28f47a05a194592b952 is no longer returned from `sub resolve()`::
```
-    return $name, $ip, $displayname, $ipcache;
+    return $name, $ip, $displayname;
```
I haven't had time to dig into it, other than checking that just reverting the single line change above isn't enough.

**Comment by drwetter** [comment]
> No -ipv6 switch supplied on the command line, so this is the desired behaviour as I understood it.

My bad. I meant that a user might expect a hint. I believe a _Message_ line would be great.

**Comment by drwetter** [comment]
> ``./program/nikto.pl  -ipv6 -h cloudflare.com -ssl  ``

Hmm, works for me now.

**Comment by drwetter** [comment]
> Hmm, works for me now. 

Small amendment (sorry). I guess that was on a different host and the IPv6 tunnel wasn't up by the time I tried.

```
prompt> ./program/nikto.pl -ipv6 -h cloudflare.com -ssl -D DS 2>&1
[..]
- Nikto v2.5.0
---------------------------------------------------------------------------
D:Mon May 17 10:16:02 2021 WARNING: No init found for nikto_core
D:Mon May 17 10:16:02 2021 'Request Hash' = {
'Connection' => 'Keep-Alive',
'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36',
'whisker' => {
'host' => 'example.com',
'ignore_duplicate_headers' => 0,
'timeout' => 10,
'include_host_in_uri' => 0,
'port' => 443,
'uri' => '/',
'force_bodysnatch' => 0,
'ssl_rsacertfile' => undef,
'ssl_save_info' => 1,
'uri_param_sep' => '?',
'method' => 'HEAD',
'MAGIC' => 31339,
'force_close' => 0,
'version' => '1.1',
'http_space2' => ' ',
'ssl_certfile' => undef,
'require_newline_after_headers' => 0,
'uri_postfix' => '',
'http_space1' => ' ',
'http_eol' => "\r\n",
'invalid_protocol_return_value' => 1,
'lowercase_incoming_headers' => 1,
'retry' => 0,
'ssl' => 1,
'force_open' => 0,
'trailing_slurp' => 0,
'keep-alive' => 1,
'max_size' => 750000,
'normalize_incoming_headers' => 1,
'protocol' => 'HTTP',
'uri_prefix' => ''
},
'Host' => 'example.com'
};
D:Mon May 17 10:16:02 2021 'Result Hash' = {
'whisker' => {
'uri' => '/',
'MAGIC' => 31340,
'error' => 'opening stream: can\'t connect (connect failed): Network is unreachable'
}
};
D:Mon May 17 10:16:02 2021 'Request Hash' = {
'Connection' => 'Keep-Alive',
'Host' => 'example.com',
'whisker' => {
[..]
'host' => 'example.com'
},
'User-Agent' => 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.83 Safari/537.36'
};
D:Mon May 17 10:16:02 2021 'Result Hash' = {
'whisker' => {
'uri' => '/',
'MAGIC' => 31340,
'error' => 'opening stream: can\'t connect (connect failed): Network is unreachable'
}
};
+ No web server found on example.com:443.
---------------------------------------------------------------------------
+ 0 host(s) tested
D:Mon May 17 10:16:02 2021 T:Mon May 17 10:16:02 2021: Ending
 ```

Maybe a better output than _No web server found on example.com:443_ would be more appropriate? Like _cloudflare.com wasn't reachable via IPv6. Please try "nitko.pl -check6" for an IPv6 connectivity check_

**Comment by sullo** [comment]
@richardleach I think removing the [] inside scrub is the best way, so that no matter how the caller sends it, it won't bomb out.

**Comment by sullo** [comment]
@drwetter I had a case where I was getting duplicate "multiple IP" messages back, such that it was like returning duplicates. I may have broken the whole thing trying to solve that. I'll review that change and see what's up. 

Also in a post you seem to be questioning why you were seeing example.com in the Host header? That's because you have -S S (scrub) which attempts to remove all the IPs/Names. Or am I missing what you were saying?

**Comment by drwetter** [comment]
The multiple IP addresses line works again, thanks! For dual stacked hosts maybe it could list also the IP addresses from the other world. E.g. if you're on perfectly connected IPv6 host and you run ``./program/nikto.pl  -h cloudflare.com -ssl``?

For the example.com issue I maybe had a case of not rtfm ;-)

**Comment by drwetter** [comment]
Thanks for all the work so far btw!!

**Comment by sullo** [comment]
> The multiple IP addresses line works again, thanks! For dual stacked hosts maybe it could list also the IP addresses from the other world. E.g. if you're on perfectly connected IPv6 host and you run `./program/nikto.pl -h cloudflare.com -ssl`?

I think that would be good in a separate issue and I'll see if I can hack it into the 2.5.0 release.

**Comment by sullo** [comment]
Can I get a final ready/not from you @richardleach  @drwetter ? Thanks

**Comment by richardleach** [comment]
I'll try to look at it later tomorrow.

**Comment by richardleach** [comment]
> > The multiple IP addresses line works again, thanks! For dual stacked hosts maybe it could list also the IP addresses from the other world. E.g. if you're on perfectly connected IPv6 host and you run `./program/nikto.pl -h cloudflare.com -ssl`?
> 
> I think that would be good in a separate issue and I'll see if I can hack it into the 2.5.0 release.

This would mean changing the `Socket::getaddrinfo` call in the `resolve` sub (in _nikto_core.plugin_). When the _-ipv4_ and _-ipv6_ switches went in, I made that pick the relevant address family.

```
my ($err, @res) = Socket::getaddrinfo($ident, "", {socktype => SOCK_RAW,
                                   family => ( $CLI{'ipv6'} ) ?  AF_INET6 :  AF_INET
                                  });
```
You'd have to take that family assignment out altogether. 

But then you have to put in some extra work to make sure that the address set in `$ip` (currently just the first value from the list: ` $ip = $addresses[0];`) is of the correct IP version. And cope when all the resolved addresses are from the unwanted IP version.

**Comment by drwetter** [comment]
Today or tommorow

**Comment by drwetter** [comment]
Basically it works. What would be great are (besides #722) some user interface issues:

1)
```
hostwithipv6|0% ./program/nikto.pl   -h ipv6.google.com
- Nikto v2.5.0
---------------------------------------------------------------------------
+ ERROR: Cannot resolve hostname 'ipv6.google.com' because 'Name or service not known'.
+ 0 host(s) tested
hostwithipv6|0% host ipv6.google.com
ipv6.google.com is an alias for ipv6.l.google.com.
ipv6.l.google.com has IPv6 address 2a00:1450:400e:807::200e
hostwithipv6|0%
```

--> misleading error message. Should be IMO ~ "Please use -ipv6 and make sure you have IPv6 connectivity ...(-check*6)"

2)
```
hostwithioutpv6|0% ./program/nikto.pl  -ipv6 -h google.com
- Nikto v2.5.0
---------------------------------------------------------------------------
+ No web server found on google.com:80.
---------------------------------------------------------------------------
+ 0 host(s) tested
```
                      
--> misleading error message. Should be IMO ~ "Cannot connect, please make sure you have IPv6 connectivity ...(-check*6)"

3)
```
hostwithtpv6|0%  ./program/nikto.pl   -h http://2a00:1450:400e:807::200e
- Nikto v2.5.0
---------------------------------------------------------------------------
+ ERROR: Cannot resolve hostname '2a00:1450:400e:807::200e' because 'Address family for hostname not supported'.
+ 0 host(s) tested
```
                
--> misleading error message. Should be IMO ~ "for IPv6 addresses please use -ipv6"

4)
```
hostwithtpv6|0%  ./program/nikto.pl   -h '[2a00:1450:400e:807::200e]'
- Nikto v2.5.0
---------------------------------------------------------------------------
+ No web server found on [2a00:1450:400e:807::200e]:80.
---------------------------------------------------------------------------
+ 0 host(s) tested
```
--> misleading error message. Should be IMO ~ "Please use -ipv6 and make sure you have IPv6 connectivity ...(-check*6)"

**Comment by sullo** [comment]
Thanks for the suggestions.
1) I don't think this will be possible, as at that point all we know is it didn't resolve and not why. The only way to figure this out would be to auto-switch to v6 and try again. I don't think we want to do that.

Updated messages for the others to the best we can in each spot.

**Comment by drwetter** [comment]
For 1): yes, maybe minor issue. Haven't looked at your resolver: Don't you have a common resolver for IPv6+IPv4? If so, one maybe can tell by the presence if an IPv6 address returned.

**Comment by richardleach** [comment]
I think this should do what @drwetter is after for 1., whilst still only scanning the requested protocol family:


```
diff --git a/program/plugins/nikto_core.plugin b/program/plugins/nikto_core.plugin
index 51fe79a..0bdea8d 100644
--- a/program/plugins/nikto_core.plugin
+++ b/program/plugins/nikto_core.plugin
@@ -1095,9 +1095,7 @@ sub resolve {

         if ( $LW2::LW2_CAN_IPv6 ) { # IPv4/v6 resolve
             use Socket qw(:addrinfo SOCK_RAW);
-            my ($err, @res) = Socket::getaddrinfo($ident, "", {socktype => SOCK_RAW,
-                                  family => ( $CLI{'ipv6'} ) ?  AF_INET6 :  AF_INET
-                                  });
+            my ($err, @res) = Socket::getaddrinfo($ident, "", {socktype => SOCK_RAW });
             if ($err) {
                 my $msg = "ERROR: Cannot resolve hostname '$ident' because '$err'.";
                 if ($ident =~ /^\[?($LW2::IPv6_re_inc_zoneid)\]?$/ ) {
@@ -1121,23 +1119,32 @@ sub resolve {
             }
         }

-        $ip = $addresses[0];
-        push(@scrub, $ip);
-        if ($addresses[1] != "") {  # have more than 1 IP
-                my %seen;
-                # get unique list of IPs
-                my @unique = grep !$seen{$_}++, @addresses;
+        my @temp4_ipcache;
+        my @temp6_ipcache;
+        my %seen;
+        foreach $temp_ip (@addresses) {
+            if ($temp_ip =~ /:/) {
+                push @temp6_ipcache, $temp_ip if !$seen{ $temp_ip }++;
+            } else {
+                push @temp4_ipcache, $temp_ip if !$seen{ $temp_ip }++;
+            }
+        }

-                for my $newip (@unique) {
-                    if ($newip ne $ip) {
-                        push(@scrub, $newip);
-                        $ipcache .= "$newip, ";
-                    }
-                }
-                if ($ipcache != "") {
-                    $ipcache =~ s/, $//;
-                    nprint("+ Multiple IP addresses found: $ip, $ipcache", "", @scrub);
-                }
+        $ip = ( $CLI{'ipv6'} ) ? shift @temp6_ipcache : shift @temp4_ipcache;
+        push(@scrub, $ip, @temp4_ipcache, @temp6_ipcache);
+        $ipcache = join ", ", (@temp4_ipcache, @temp6_ipcache);
+
+        if ( $ip eq '' ) {
+            if ( $CLI{'ipv6'} && scalar @temp4_ipcache ) {
+                nprint("+ ERROR: IPv6 scanning mode requested but only IPv4 addresses found ($ipcache)");
+            } elsif ( $CLI{'ipv4'} && scalar @temp6_ipcache ) {
+                nprint("+ ERROR: IPv4 scanning mode requested but only IPv6 addresses found ($ipcache)");
+            }
+            exit;
+        }
+
+        if ($ipcache != "") {
+            nprint("+ Multiple IP addresses found: $ip, $ipcache", "", @scrub);
         }

         if ( $ip !~ /^$LW2::IPv4_re$/ &&
```

I've only been able to snatch a small amount of time to write & test it though.

**Comment by richardleach** [comment]
I've updated the local copy of the _nikto-2.5.0_ branch on my EC2 instance and it works okay:
```
$ perl ./nikto.pl -check6
Performing IPv6 connectivity tests:
+ This version of Socket (2.027) does support IPv6
+ Successful DNS resolution of 'ipv6.google.com': 2a00:1450:4009:820::200e
+ Successful TCP connection to 'ipv6.google.com:443'
----> All tests successful
```

```
$ ./nikto.pl -h ipv6.google.com
- Nikto v2.5.0
---------------------------------------------------------------------------
+ ERROR: IPv4 scanning mode requested but only IPv6 addresses found (2a00:1450:4009:820::200e)
```

```
$ ./nikto.pl -ipv6 -h ipv6.google.com
- Nikto v2.5.0
---------------------------------------------------------------------------
+ Target IP:          2a00:1450:4009:820::200e
+ Target Hostname:    ipv6.google.com
+ Target Port:        80
+ Start Time:         2021-06-09 19:45:01 (GMT0)
---------------------------------------------------------------------------
+ Server: gws
+ Uncommon header 'bfcache-opt-in' found, with contents: unload.
+ The X-Content-Type-Options header is not set. This could allow the user agent to render the content of the site in a different fashion to the MIME type. See: https://www.netsparker.com/web-vulnerability-scanner/vulnerabilities/missing-content-type-header/
+ Root page / redirects to: https://ipv6.google.com/?gws_rd=ssl
...etc...
```
Firing up wireshark might be the simplest way to check if _host_ and _Socket_ are sending similar packets to the same nameservers or not.

**Comment by richardleach** [comment]
The [Socket docs](https://metacpan.org/pod/Socket::GetAddrInfo) mention failures on older Darwin machines (e.g. osvers=8.11.1), if that's relevant.

**Comment by sullo** [comment]
I'm not as concerned about my setup working perfectly (though it'd be nice for testing) as long as it works for most setups.

What I was trying to work on & test was what happens if a host resolves both 6 and 4? Ideally I'd like to see the multiple IP addresses report both. So for example on google.com:

```
sullo@talyn: ~/dev/nikto/program# host google.com
google.com has address 172.217.12.238
google.com has IPv6 address 2607:f8b0:4004:806::200e
``` 
We would see something like:
`+ Message:            Multiple IP addresses found: 172.217.12.238, 2607:f8b0:4004:806::200e`

I'm unclear if this would happen now, but looking at the code I don't think it will.

**Comment by richardleach** [comment]
It should report both, e.g. from my EC2 instance:
```
$ ./nikto.pl -h google.com
- Nikto v2.5.0
---------------------------------------------------------------------------
+ Multiple IP addresses found: 172.217.169.78, 2a00:1450:4009:81d::200e
+ Target IP:          172.217.169.78
+ Target Hostname:    google.com
+ Target Port:        80
+ Start Time:         2021-06-10 20:43:41 (GMT0)
---------------------------------------------------------------------------


$ ./nikto.pl -h bbc.co.uk
- Nikto v2.5.0
---------------------------------------------------------------------------
+ Multiple IP addresses found: 151.101.192.81, 151.101.0.81, 151.101.64.81, 151.101.128.81, 2a04:4e42:200::81, 2a04:4e42:400::81, 2a04:4e42:600::81, 2a04:4e42::81
+ Target IP:          151.101.192.81
+ Target Hostname:    bbc.co.uk
+ Target Port:        80
+ Start Time:         2021-06-10 20:44:38 (GMT0)
---------------------------------------------------------------------------
```

**Comment by sullo** [comment]
Thanks @richardleach! 

I feel like this is ready, @drwetter & @richardleach do you concur?

**Comment by richardleach** [comment]
It seems to be working fine here, so unless there are any other reports of problems or change requests, it's a thumbs up from me. 

(Though I'd still like to know what's the cause of your failed IPv6 DNS lookups when _host_ can manage it.)

**Comment by drwetter** [comment]
sh... totally forgot this. Will do later during the week

**Comment by drwetter** [comment]
2. is still the same for me, other than that it looks good to me

**Comment by sullo** [comment]
@drwetter for your 2 issue is this better, or not what you're seeing?

```
 ~/dev/nikto/program# ./nikto.pl -ipv6 -h googleaaaad.com
- Nikto v2.5.0
---------------------------------------------------------------------------
+ ERROR: Cannot resolve hostname 'googleaaaad.com' because 'nodename nor servname provided, or not known'. Ensure you have IPv6 connectivity. Trying running Nikto with the '-check6' flag.
+ 0 host(s) tested
```

**Comment by drwetter** [comment]
Looks good to me, original example:

```
prompt> ./nikto.pl -ipv6 -h google.com 
- Nikto v2.5.0
---------------------------------------------------------------------------
+ Multiple IP addresses found: 2a00:1450:4001:80f::200e, 142.250.185.110
+ Unable to connect to google.com:80. Ensure you have IPv6 connectivity. Trying running Nikto with the '-check6' flag.
---------------------------------------------------------------------------
+ 0 host(s) tested
prompt>
```

**Comment by sullo** [comment]
Based on the feedback and my own testing (after I fixed v6 on my router) I'm going to close this as complete.

Thank you @drwetter and @richardleach for the code and testing! I really appreciate the contributions.
